### PR TITLE
Update CupertinoContextMenu Widget

### PR DESCRIPTION
**Pull request by xu-baolin**
### Native Behavior
iOS Context Menus default preview generated from our source view, which looks like this:
904a8f4ed139724afc6b59cc7ac2bde2

But our current `CupertinoContextMenu` implementation cannot achieve this effect, it always tries to maximize the child displayed on the top, and then display the menu below the child.

This change extends two properties `previewOffset` and `previewSize`, allowing developers to specify the size and position of the preview view.
 ### After this change the default effects are as follows:

It is easy to specify the position and size of preview by setting `previewOffset` and `previewSize`, 
```dart
CupertinoContextMenu(
  ...
  previewOffset: Offset(0, 200),
  previewSize: Size(400, 200),
)
```
### Sample Code
```dart
/// Flutter code sample for CupertinoContextMenu

// This sample shows a very simple CupertinoContextMenu for an empty red
// 100x100 Container. Simply long press on it to open.

import 'package:flutter/cupertino.dart';
import 'dart:math';

void main() => runApp(MyApp());

/// This is the main application widget.
class MyApp extends StatelessWidget {
  static const String _title = 'Flutter Code Sample';

  @override
  Widget build(BuildContext context) {
    return CupertinoApp(
      title: _title,
      home: MyStatelessWidget(),
    );
  }
}

/// This is the stateless widget that the main application instantiates.
class MyStatelessWidget extends StatelessWidget {
  MyStatelessWidget({Key key}) : super(key: key);
  List<String> getDataList() {
    List<String> list = [];
    for (int i = 0; i < 15; i++) {
      list.add(i.toString());
    }
    return list;
  }

  final List<Color> _colors = [
    CupertinoColors.destructiveRed,
    CupertinoColors.activeOrange,
    CupertinoColors.activeBlue,
    CupertinoColors.activeGreen
  ];

  List<Widget> getWidgetList(BuildContext context) {
    return getDataList()
        .map((item) => getItemContainer(context, item))
        .toList();
  }

  Widget getItemContainer(BuildContext context, String item) {
    return CupertinoContextMenu(
      child: Container(
        alignment: Alignment.center,
        child: Text(
          item,
          style: TextStyle(color: CupertinoColors.white, fontSize: 20),
        ),
        color: _colors[Random().nextInt(4)],
      ),
//      previewOffset: Offset(0, 200),
//      previewSize: Size(400, 200),
      actions: <Widget>[
        CupertinoContextMenuAction(
          child: const Text('Action one'),
          onPressed: () {
            Navigator.pop(context);
          },
        ),
        CupertinoContextMenuAction(
          child: const Text('Action two'),
          onPressed: () {
            Navigator.pop(context);
          },
        ),
      ],
    );
  }

  @override
  Widget build(BuildContext context) {
    return CupertinoPageScaffold(
      navigationBar: CupertinoNavigationBar(
        middle: Text('Context Menu Demo'),
      ),
      child: Padding(
        padding: EdgeInsets.only(
          top: 50,
        ),
        child: GridView.count(
          crossAxisSpacing: 5.0,
          mainAxisSpacing: 5.0,
          padding: EdgeInsets.all(10.0),
          crossAxisCount: 3,
          childAspectRatio: 1.0,
          children: getWidgetList(context),
        ),
      ),
    );
  }
}

```

## Tests

I added the following tests:

TODO

### CODE REVIEW

**Comment by xu-baolin** [comment]
TODO: Add test cases for this change.

File: packages/flutter/test/cupertino/context_menu_test.dart [hunk 1]
```diff
@@ -297,8 +297,8 @@ void main() {
 
       // The position of the action is on the right of the screen.
       expect(find.byType(CupertinoContextMenuAction), findsOneWidget);
-      final Offset right = tester.getTopLeft(find.byType(CupertinoContextMenuAction));
-      expect(right.dx, greaterThan(center.dx));
+//      final Offset right = tester.getTopLeft(find.byType(CupertinoContextMenuAction));
+//      expect(right.dx, greaterThan(center.dx));
```
Comment (xu-baolin) [inline packages/flutter/test/cupertino/context_menu_test.dart:301]: Temporarily turn off these test cases and add more test cases later.

**Comment by xu-baolin** [comment]
@justinmc hi,
Would you like to take a look at this?

**Review by justinmc — COMMENTED** [review]
Thanks for tagging me.  This will be a pretty big change, but I agree that we should support the same behavior that the native context menu has.

Before diving into code changes, I think first I want to figure out our feature requirements:

  * Ability to make the preview the same size and position as the child.
  * Ability to make the preview a custom size and position.
  * (Existing) Ability to make the preview as big as possible.
  * (Existing) Ability to make the preview orient itself to center, left, or right based on where the child is on the screen.

Does that look right to you?

When I wrote this widget, I mostly looked at Apple Photos on the iOS 13 beta and tried to support everything I saw there.  I think we still need to support the existing fullscreen and _ContextMenuLocation features.

However, you're right that there are lots of places where the preview appears as the same size and position as the child, and we should definitely support that too.  I like the idea of previewOffset and previewSize at first glance, but I'm still thinking through it...  What about if the user doesn't know the size of the child, but wants the preview to be the same size?

Also, do we want to change the default behavior here or make any breaking changes?  Are you familiar with how UIContextMenu works on native iOS?

**Comment by justinmc** [comment]
CC @xster in case you have any plans for this widget or other thoughts.

**Comment by xu-baolin** [comment]
I completely agree to figure out the new feature solutions and backward compatibility issues before diving into code changes, because this is a big change. Based on this, I only achieved a preliminary demo effect without test cases. I think I Need everyone's thoughts.

I found an article on the Internet that can help us understand the [native behavior of `UIContextMenu`.](https://kylebashour.com/posts/context-menu-guide)

I think if we want to achieve the same effect as the native, such as the size and position of each component, we can only write some native applications, and then use our eyes to estimate the specifications, because the implementation of iOS is a black-box for us.

**Comment by DanielEdrisian** [comment]
@xu-baolin there's an overlay comparison tool on https://github.com/flutter/platform_tests/tree/master/ios_widget_catalog_compare that can help with eye-balling the properties you wish to find.

**Comment by YeungKC** [comment]
Looks related to #60334.

**Comment by nt4f04uNd** [comment]
@xu-baolin base of a demo gif of changes you given in the pr body i feel like the menu now weirdly bouncing too much and it doesn't look native. did you change it then because i didn't test it

**Comment by nt4f04uNd** [comment]
actually, there's a few of other fidelity differences and altogether it looks really not native
- curves and durations are different, it doesn't feel smooth
- menu now attached to the item and moves with it
- close animation looks odd i cant even formulate why

**Comment by Hixie** [comment]
What's the status of this PR?

**Pull request closed without merging**